### PR TITLE
interfaces/builtin: add kvm interface

### PR DESCRIPTION
**Pull request by Saviq**
Read/write access to /dev/kvm for kvm virtual machines

### CODE REVIEW

File: interfaces/builtin/kvm.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  kvm:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const kvmConnectedPlugAppArmor = `
+# Description: Allow write access to kvm.
+# See 'man 4 mem' for details.
```
Comment (mvo5) [inline interfaces/builtin/kvm.go]: When I open this man-page and search for "kvm" I don't find anything. What am I missing?

Comment (Saviq) [inline interfaces/builtin/kvm.go]: Of course, this was a remnant of the interface I based this on.
Fixed.

File: interfaces/builtin/kvm.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+// Getter for the name of the kvm interface
+func (iface *kvmInterface) Name() string {
+	return "kvm"
+}
+
+func (iface *kvmInterface) MetaData() interfaces.MetaData {
+	return interfaces.MetaData{
+		Summary:              kvmSummary,
+		ImplicitOnClassic:    true,
```
Comment (zyga) [inline interfaces/builtin/kvm.go]: I think this could be implicit on core as well

File: interfaces/builtin/kvm.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	}
+}
+
+func (iface *kvmInterface) String() string {
+	return iface.Name()
+}
+
+// Check validity of the defined slot
+func (iface *kvmInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	// Does it have right type?
```
Comment (zyga) [inline interfaces/builtin/kvm.go]: I'll tweak this to look nicer, I have a helper I wanted to land everywhere.

File: interfaces/builtin/kvm.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	if !(slot.Snap.Type == "os") {
+		return fmt.Errorf("%s slots only allowed on core snap", iface.Name())
+	}
+	return nil
+}
+
+// Checks and possibly modifies a plug
+func (iface *kvmInterface) SanitizePlug(plug *interfaces.Plug) error {
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface))
```
Comment (zyga) [inline interfaces/builtin/kvm.go]: `iface.Name()`

File: interfaces/builtin/kvm.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+// Check validity of the defined slot
+func (iface *kvmInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	// Does it have right type?
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	// Creation of the slot of this type
+	// is allowed only by a gadget or os snap
+	if !(slot.Snap.Type == "os") {
```
Comment (niemeyer) [inline interfaces/builtin/kvm.go]: Please use the constant here: snap.TypeOS

File: interfaces/builtin/kvm.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+// Check validity of the defined slot
+func (iface *kvmInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	// Does it have right type?
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	// Creation of the slot of this type
+	// is allowed only by a gadget or os snap
```
Comment (niemeyer) [inline interfaces/builtin/kvm.go]: Not on the gadget.

File: interfaces/builtin/kvm.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/udev"
+)
+
+const kvmSummary = `allows access to kvm`
```
Comment (niemeyer) [inline interfaces/builtin/kvm.go]: "allows access to the kvm device"

So people don't expect to get access to binaries as well. We may go there in the future, but that's not what the interface is currently doing.

File: interfaces/builtin/kvm_test.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+	c.Assert(func() { s.iface.SanitizeSlot(&interfaces.Slot{SlotInfo: &snap.SlotInfo{Interface: "other"}}) },
+		PanicMatches, `slot is not of interface "kvm"`)
+	c.Assert(func() { s.iface.SanitizePlug(&interfaces.Plug{PlugInfo: &snap.PlugInfo{Interface: "other"}}) },
+		PanicMatches, `plug is not of interface "kvm"`)
+}
+
+func (s *kvmInterfaceSuite) TestUsedSecuritySystems(c *C) {
+	expectedSnippet1 := `
+# Description: Allow write access to kvm.
+# See 'man 4 mem' for details.
```
Comment (jdstrand) [inline interfaces/builtin/kvm_test.go]: Here you reference 'man 4 mem' but up above you reference 'man kvm'. As such, this is why the testsuite is failing.

**Comment by Saviq** [comment]
All requested changes applied.

**Comment by mvo5** [comment]
unit tests are failing with: src/github.com/snapcore/snapd/interfaces/builtin/kvm.go:56: undefined: interfaces.MetaData

**Pull request merged**